akka-sample-distributed-data-scala/src/main/scala/sample/distributeddata/ReplicatedMetrics.scala

```diff
@@ -36,80 +41,104 @@ object ReplicatedMetrics {
-
-  val tickTask = context.system.scheduler.schedule(measureInterval, measureInterval,
-    self, Tick)(context.dispatcher)
-  val cleanupTask = context.system.scheduler.schedule(cleanupInterval, cleanupInterval,
-    self, Cleanup)(context.dispatcher)
-  val memoryMBean: MemoryMXBean = ManagementFactory.getMemoryMXBean
-
-  val UsedHeapKey = LWWMapKey[String, Long]("usedHeap")
-  val MaxHeapKey = LWWMapKey[String, Long]("maxHeap")
-
-  replicator ! Subscribe(UsedHeapKey, self)
-  replicator ! Subscribe(MaxHeapKey, self)
-
-  cluster.subscribe(self, InitialStateAsEvents, classOf[MemberUp], classOf[MemberRemoved])
-
-  override def postStop(): Unit = {
-    tickTask.cancel()
-    cluster.unsubscribe(self)
-    super.postStop()
-  }
-
-  var maxHeap = Map.empty[String, Long]
-  var nodesInCluster = Set.empty[String]
-
-  def receive = {
-    case Tick =>
-      val heap = memoryMBean.getHeapMemoryUsage
-      val used = heap.getUsed
-      val max = heap.getMax
-      replicator ! Update(UsedHeapKey, LWWMap.empty[String, Long], WriteLocal)(_ + (node -> used))
-      replicator ! Update(MaxHeapKey, LWWMap.empty[String, Long], WriteLocal) { data =>
-        data.get(node) match {
-          case Some(`max`) => data // unchanged
-          case _           => data + (node -> max)
+  def apply: Behavior[Command] = apply(1.second, 1.minute)
+
+  def apply(measureInterval: FiniteDuration, cleanupInterval: FiniteDuration): Behavior[Command] =
+    Behaviors.setup { context =>
+      DistributedData.withReplicatorMessageAdapter[Command, LWWMap[String, Long]] { replicator =>
+        Behaviors.withTimers[Command] { timers =>
+          implicit val selfUniqueAddress: SelfUniqueAddress =
+            DistributedData(context.system).selfUniqueAddress
+          implicit val cluster = Cluster(context.system)
+          val node = nodeKey(cluster.selfMember.address)
+
+          timers.startTimerWithFixedDelay(Tick, Tick, measureInterval)
+          timers.startTimerWithFixedDelay(Cleanup, Cleanup, cleanupInterval)
+          val memoryMBean: MemoryMXBean = ManagementFactory.getMemoryMXBean
+
+          val UsedHeapKey = LWWMapKey[String, Long]("usedHeap")
+          val MaxHeapKey = LWWMapKey[String, Long]("maxHeap")
+
+          replicator.subscribe(UsedHeapKey, InternalSubscribeResponse.apply)
+          replicator.subscribe(MaxHeapKey, InternalSubscribeResponse.apply)
+
+          val memberUpRef   = context.messageAdapter(InternalClusterMemberUp.apply)
+          val memberDownRef = context.messageAdapter(InternalClusterMemberRemoved.apply)
+          cluster.subscriptions ! Subscribe(memberUpRef, classOf[ClusterEvent.MemberUp])
+          cluster.subscriptions ! Subscribe(memberDownRef, classOf[ClusterEvent.MemberRemoved])
+
+          def postStop(): Unit = {
+            timers.cancel(Tick)
+            cluster.subscriptions ! Unsubscribe(memberUpRef)
+            cluster.subscriptions ! Unsubscribe(memberDownRef)
```

Comment:
these postStop things are not needed, handled automatically when the actor stops